app/controllers/api/storage.php

```diff
@@ -928,41 +1043,95 @@
 
         $path = $file->getAttribute('path', '');
 
-        if (!\file_exists($path)) {
-            throw new Exception('File not found in '.$path, 404);
+        $device = Storage::getDevice('files');
+        
+        if (!$device->exists($path)) {
+            throw new Exception('File not found in ' . $path, 404);
         }
 
-        $device = Storage::getDevice('files');
+        $usage
+            ->setParam('storage.files.read', 1)
+            ->setParam('bucketId', $bucketId)
+        ;
 
-        $source = $device->read($path);
+        $response
+            ->setContentType($file->getAttribute('mimeType'))
+            ->addHeader('Expires', \date('D, d M Y H:i:s', \time() + (60 * 60 * 24 * 45)) . ' GMT') // 45 days cache
+            ->addHeader('X-Peak', \memory_get_peak_usage())
+            ->addHeader('Content-Disposition', 'attachment; filename="' . $file->getAttribute('name', '') . '"')
+        ;
+
+        $size = $file->getAttribute('sizeOriginal', 0);
+
+        $rangeHeader = $request->getHeader('range');
+        if(!empty($rangeHeader)) {  
+            list($unit, $range) = explode('=', $rangeHeader);
+            if($unit == 'bytes' && !empty($range)) {
+                list($rangeStart, $rangeEnd) = explode('-', $range);
+                if(strlen($rangeStart) == 0 || strstr($range, '-') === false) {
+                    throw new Exception('Invalid range', 416);
+                }
+                $rangeStart = (int) $rangeStart;
+                if(strlen($rangeEnd) == 0) {
+                    $rangeEnd =  min(($rangeStart + 2000000-1), ($size - 1));
+                } else {
+                    $rangeEnd = (int) $rangeEnd;
+                }
+                if(($rangeStart >= $rangeEnd) || $rangeEnd >= $size) {
+                    throw new Exception('Invalid range', 416);
+                }
```

Comment:
Is it possible to make part of this in the `utopia-php` library?

I feel like we would need to do this in every endpoint with chunked downloads 🤔

Reply:
may be add to framework, get range header or something?